### PR TITLE
[SYNPY-1320] Upload benchmark + Documentation

### PR DESCRIPTION
**Pull request by BryanFauble**
Problem:

1. OTEL context propagation was broken due to the custom threading in the sync functions
2. We did not have upload benchmarks documented
3. Documentation in the README was lacking for programmatic access

Solution:

1. Propagate the OTEL context into the `syn.store` function
2. Create documentation in the benchmarking rst
3. Add some examples to the README that talk over some common use-cases for file/folder/project creation

Testing:

1. I ran and collected the benchmark information on EC2

### CODE REVIEW

File: synapseclient/client.py [hunk 1]
```diff
@@ -1274,253 +1275,266 @@ def store(
...
-                trace.get_current_span().set_attributes({"synapse.id": ""})
-                return type(obj)(**self.restPOST(obj.postURI(), obj.json()))
-
-            except SynapseHTTPError as err:
-                # If already present and we want to update attempt to get the object content
-                if createOrUpdate and err.response.status_code == 409:
-                    newObj = self.restGET(obj.getByNameURI(obj.name))
-                    newObj.update(obj)
-                    obj = type(obj)(**newObj)
+        with tracer.start_as_current_span(
```
Comment (BryanFauble) [inline synapseclient/client.py:1278]: Adding this `with` is the only real change here - The sync functions multi-thread at this point and OTEL context was not propagating

File: README.md [hunk 2]
```diff
@@ -87,35 +87,137 @@ Usage as a library
...
+import synapseclient
 
-    syn = synapseclient.Synapse()
+syn = synapseclient.Synapse()
+
+## log in using auth token
+syn.login(authToken='auth_token')
+```
+
+#### Create a manifest TSV and sync a directory to synapse
```
Comment (thomasyu888) [inline README.md:102]: Nit: I think it's great this is in the readme, but eventually, we should have an article with this sort of information.

Comment (thomasyu888) [inline README.md:102]: NIt: not everyone will know what `manifest` means, so we could potentially simplify this with `sync a local directory to Synapse`.

File: synapseclient/client.py (hunk 1 above)
Comment (danlu1) [reply synapseclient/client.py:1278]: Do you mind adding the explanation of opentelemetry_context parameter in the function docstring?

File: docs/articles/benchmarking.rst [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+| 25 Files 1MB total size   | 10.43s            | 8.99s          | 1.83s   |
++---------------------------+-------------------+----------------+---------+
+| 775 Files 10MB total size | 243.57s           | 257.27s        | 7.64s   |
++---------------------------+-------------------+----------------+---------+
+| 10 Files 1GB total size   | 27.18s            | 33.73s         | 16.31s  |
++---------------------------+-------------------+----------------+---------+
+| 10 Files 100GB total size | 3211s             | 3047s          | 3245s   |
++---------------------------+-------------------+----------------+---------+
+
+Benchmarking Script::
```
Comment (thomasyu888) [inline docs/articles/benchmarking.rst:33]: Should we save this script in `./docs/scripts/`

File: setup.cfg [hunk 4]
```diff
@@ -51,9 +51,9 @@ install_requires =
     keyring>=15,<23.5
     keyrings.alt==3.1; sys_platform == "linux"
     deprecated>=1.2.4,<2.0
-    opentelemetry-api~=1.20.0
-    opentelemetry-sdk~=1.20.0
-    opentelemetry-exporter-otlp-proto-http~=1.20.0
+    opentelemetry-api>=1.20.0
```
Comment (thomasyu888) [inline setup.cfg:54]: `>`?  Should we lock create a `<` so that we don't run into potential dependency hell?

**Review by thomasyu888 — APPROVED** [review]
🔥 Looks great to me!  I'm going to pre-approve.  

I do realize we've baked in otel contexts throughout the code which adds a layer of nesting sometimes which is ok.  In the beginning, I was wondering if there was a way for that code to exist without the opentel packages installed (so we don't always require the dependency) but it should be fine.

File: README.md (hunk 2 above)
Comment (BryanFauble) [reply README.md:102]: I'm think of dropping the 'many files' part in lieu of "more than one"

File: docs/articles/benchmarking.rst [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+The results were created on a `t3a.micro` EC2 instance with a 200GB disk size running in us-east-1.
+The script that was run can be found below.
+
+
++---------------------------+-------------------+----------------+---------+
+| Test                      | Synapseutils Sync | Manual os.walk | S3 Sync |
++===========================+===================+================+=========+
+| 25 Files 1MB total size   | 10.43s            | 8.99s          | 1.83s   |
++---------------------------+-------------------+----------------+---------+
+| 775 Files 10MB total size | 243.57s           | 257.27s        | 7.64s   |
```
[the remark this answers is not in the document]
Comment (BryanFauble) [reply docs/articles/benchmarking.rst:26]: > Are you including the time you are taking to create the folders?

This is not included in the time.

File: README.md (hunk 2 above)
Comment (thomasyu888) [reply README.md:102]: Got it.  Looking at your benchmarking, i did notice that there are times where `sync` is not as optimal of a method, but we can still just recommend this approach.

Heres an issue we should resolve: https://sagebionetworks.jira.com/browse/SYNPY-1179.

File: setup.cfg (hunk 4 above)
Comment (BryanFauble) [reply setup.cfg:54]: I am going to go back to `~=1.21.0`

File: docs/articles/benchmarking.rst (hunk 3 above)
Comment (BryanFauble) [reply docs/articles/benchmarking.rst:33]: Good idea - I missed that directory

File: README.md (hunk 2 above)
Comment (BryanFauble) [reply README.md:102]: > Nit: I think it's great this is in the readme, but eventually, we should have an article with this sort of information.

I agree - I am going to make a ticket at the top of the backlog to absorb this kind of information into https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/docs/getting_started/basics.rst

File: README.md [hunk 6]
```diff
@@ -85,37 +85,139 @@ Note that a [Synapse account](https://www.synapse.org/#RegisterAccount:0) is req
...
 
-    syn = synapseclient.Synapse()
+syn = synapseclient.Synapse()
+
+## log in using auth token
+syn.login(authToken='auth_token')
+```
+
+#### Sync a local directory to synapse
+This is the recommended way of synchronizing more than one file or directory to a synapse project is through the use of `synapseutils`. Using this library let's us handle scheduling everything required to sync an entire directory tree. Read more about the manifest file format in [`synapseutils.syncToSynapse`](https://python-docs.synapse.org/build/html/articles/synapseutils.html#synapseutils.sync.syncToSynapse)
```
Comment (danlu1) [inline README.md:103]: Nit: could you rephrase this sentence? It has two verbs.

**Review by danlu1 — APPROVED** [review]

**Comment by danlu1** [comment]
Thanks for working on this. I approved the PR.

File: README.md (hunk 6 above)
Comment (BryanFauble) [reply README.md]: I modified this slightly - Let me know if this new version sounds better:

> This is the recommended way of synchronizing more than one file or directory to a synapse project through the use of `synapseutils`. Using this library allows us to handle scheduling everything required to sync an entire directory tree. Read more about the manifest file format in [`synapseutils.syncToSynapse`](https://python-docs.synapse.org/build/html/articles/synapseutils.html#synapseutils.sync.syncToSynapse)

Comment (danlu1) [reply README.md]: Looks great!
